File: frontend/src/api/train.ts
import api from './axios'

export interface TrainJobRequest {
  dataset_id: string
  version?: string
  model_name?: string
  epochs: number
  imgsz?: number
  batch?: number
  base_model_id?: string  // 用于微调的已有模型ID
  business?: string       // 业务/算法类型（守门员按此隔离对比），默认 "general"
  // 阶段1.3（训练任务聚合）：多数据集聚合训练
  dataset_ids?: string[]      // 待聚合训练的数据集列表（缺省=仅 dataset_id）
  aggregate_incomplete?: boolean // 阶段1.4（雪球）：自动带同业务未完成训练数据集
  // 高级训练参数（可选，缺省则由系统默认）
  lr0?: number          // 初始学习率
  optimizer?: string    // auto / SGD / Adam / AdamW
  weight_decay?: number // 权重衰减
  patience?: number     // 早停轮数（0 = 关闭）
  gpu_index?: number | null // 训练节点：指定 GPU 索引（null/undefined = 自动）
  // 训练增强（1.5/1.6）：样本池抽样 + 回忆集混训
  hard_sample_ratio?: number      // 困难样本库按比例抽样并入训练（0=关闭，默认 0.1）
  background_sample_ratio?: number // 空白样本库抽样（负样本，默认 0.05）
  recall_enabled?: boolean        // 回忆集混训：增量训练时混入同业务旧数据防遗忘（默认 true）
}

export interface GpuInfo {
  index: number
  name: string
  total_gb: number
  used_gb: number
  free_gb: number
}

export interface GpuListResult {
  cuda_available: boolean
  gpus: GpuInfo[]
}

export interface LabelFilterInfo {
  model_code?: string
  kept: string[]
  dropped: string[]
}

export interface CreateTrainJobResult {
  job_id: string
  status: string
  label_filter?: LabelFilterInfo
  dataset_ids?: string[]
  aggregated?: boolean
  aggregation?: {
    dataset_count: number
    unified_names: string[]
    total_images: number
  }
}

export const createTrainJob = async (params: TrainJobRequest): Promise<CreateTrainJobResult> => {
  const { data } = await api.post('/train/jobs', params)
  return data
}

/**
 * 训练前确保预训练权重已本地化（镜像加速下载缓存），避免启动阶段联网下载卡顿。
 * 已缓存时立即返回（status=ready），自定义/微调模型同样秒返。
 */
export const prepareWeights = async (model_name: string) => {
  const { data } = await api.post('/train/prepare-weights', { model_name })
  return data
}

/**
 * 获取当前服务器的 GPU 列表（训练节点），无 GPU 时 cuda_available=false
 */
export const listGpus = async (): Promise<GpuListResult> => {
  const { data } = await api.get('/train/gpus')
  return data
}

// 根据数据类别名自动推断业务/算法类型（无需手动选择业务场景）
export const inferBusiness = async (dataset_id: string, version: string = 'v1') => {
  const { data } = await api.get('/train/business', {
    params: { dataset_id, version }
  })
  return data
}

/**
 * 根据当前硬件与数据集自动推荐训练参数（基础 + 高级，可手动覆盖）
 */
export const suggestTrainParams = async (
  dataset_id: string,
  version: string = 'v1',
  base_model_id?: string
) => {
  const { data } = await api.get('/train/suggest', {
    params: { dataset_id, version, base_model_id: base_model_id || undefined }
  })
  return data
}

export const listTrainJobs = async () => {
  const { data } = await api.get('/train/jobs')
  return data
}

export const getTrainJob = async (jobId: string) => {
  const { data } = await api.get(`/train/jobs/${jobId}`)
  return data
}

export const getTrainJobTree = async (jobId: string) => {
  const { data } = await api.get(`/train/jobs/${jobId}/tree`)
  return data
}

export const stopTrainJob = async (jobId: string) => {
  const { data } = await api.post(`/train/jobs/${jobId}/stop`)
  return data
}

export const resumeTrainJob = async (jobId: string) => {
  const { data } = await api.post(`/train/jobs/${jobId}/resume`)
  return data
}

export const deleteTrainJob = async (jobId: string) => {
  const { data } = await api.delete(`/train/jobs/${jobId}`)
  return data
}

File: backend/src/api/routes/train.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from src.services.train_service import TrainService

router = APIRouter(prefix="/train", tags=["train"])
train_service = TrainService()

class TrainJobRequest(BaseModel):
    dataset_id: str
    version: str = "v1"
    model_name: str = "yolov8n.pt"  # 可以是预训练模型或已有模型ID
    epochs: int = 10
    imgsz: int = 640
    batch: int = -1
    base_model_id: Optional[str] = None  # 用于微调的已有模型ID
    business: Optional[str] = None       # 业务/算法类型：传空则系统按数据集类别名自动分配（守门员按此隔离对比）
    # 高级训练参数（可选，缺省则由系统默认）
    lr0: Optional[float] = None          # 初始学习率
    optimizer: Optional[str] = None      # auto / SGD / Adam / AdamW
    weight_decay: Optional[float] = None # 权重衰减
    patience: Optional[int] = None       # 早停轮数（0 = 关闭）
    gpu_index: Optional[int] = None      # 训练节点：指定 GPU 索引（多卡环境；缺省自动）

@router.get("/gpus")
async def list_gpus():
    """列出当前服务器可用的 GPU（训练节点），用于创建任务时选择"""
    return await train_service.list_gpus()

@router.get("/suggest")
async def suggest_train_params(dataset_id: str, version: str = "v1", base_model_id: str = None):
    """自动推荐训练参数（基础 + 高级），根据硬件与数据集规模"""
    return await train_service.suggest_params(dataset_id, version, base_model_id)

@router.get("/business")
async def infer_business(dataset_id: str, version: str = "v1"):
    """根据数据集类别名自动推断业务/算法类型（选中数据集后系统自动分配，无需手动选择）"""
    return await train_service.infer_dataset_business(dataset_id, version)

@router.post("/jobs")
async def create_train_job(request: TrainJobRequest):
    """创建训练任务（支持基于已有模型微调）"""
    try:
        result = await train_service.create_job(
            dataset_id=request.dataset_id,
            version=request.version,
            model_name=request.model_name,
            epochs=request.epochs,
            imgsz=request.imgsz,
            batch=request.batch,
            base_model_id=request.base_model_id,
            business=request.business,
            lr0=request.lr0,
            optimizer=request.optimizer,
            weight_decay=request.weight_decay,
            patience=request.patience,
            gpu_index=request.gpu_index
        )
        return result
    except ValueError as e:
        raise HTTPException(400, str(e))

@router.get("/jobs")
async def list_train_jobs():
    """列出所有训练任务"""
    return await train_service.list_jobs()

@router.get("/jobs/{job_id}")
async def get_train_job(job_id: str):
    """获取训练任务详情"""
    result = await train_service.get_job(job_id)
    if not result:
        raise HTTPException(404, "Job not found")
    return result

@router.get("/jobs/{job_id}/tree")
async def get_train_job_tree(job_id: str):
    """获取训练任务输出目录树（文件夹结构）"""
    result = await train_service.get_job_tree(job_id)
    if not result:
        raise HTTPException(404, "Job not found")
    return result

@router.post("/jobs/{job_id}/stop")
async def stop_train_job(job_id: str):
    """停止训练任务"""
    result = await train_service.stop_job(job_id)
    return result

@router.post("/jobs/{job_id}/resume")
async def resume_train_job(job_id: str):
    """继续训练中断的任务（支持正常停止和崩溃恢复）"""
    try:
        result = await train_service.resume_job(job_id)
        if not result:
            raise HTTPException(404, "Job not found")
        return result
    except ValueError as e:
        raise HTTPException(400, str(e))
    except Exception as e:
        raise HTTPException(500, f"Failed to resume training: {str(e)}")

@router.delete("/jobs/{job_id}")
async def delete_train_job(job_id: str):
    """删除训练任务"""
    result = await train_service.delete_job(job_id)
    if not result:
        raise HTTPException(404, "Job not found")
    return result
